### ISSUE TITLE
Why to collect all examples from all iterations?

### ISSUE THREAD

**Issue by evg-tyurin** [suragnair/alpha-zero-general#32]
https://github.com/suragnair/alpha-zero-general/blob/38fc2983a1dd85104348d72420284daa6a2df469/Coach.py#L69

What is the main goal of collecting all examples from all iterations? On one hand we get more smooth loss dynamics and more examples for the same computation cost, on other hand when we are far enough from the first iterations our model has gained some experience and old examples depict weak predictions and thus pull the model back.

**Comment by suragnair** [comment]
In the paper on AlphaGo Zero they play 25000 games in each iteration. Quoting the paper, "Each mini­batch of data is sampled uniformly at random from all positions of the most recent 500,000 games of self­play." There is no real explanation provided in the paper. 

I do agree with you that it seems counterintuitive to use old examples. However the old examples are only from the last 20 iterations in AlphaGo Zero. If training for long enough, having more examples probably provides a richer and diverse training set.

**Comment by evg-tyurin** [comment]
OK, thank you for the explanation.

What do you think, if we plan long training and want to keep the above approach without having the program runs continuously then we should backup appropriate train examples before stop/restart the program.  Isn't it?

**Comment by suragnair** [comment]
Sounds reasonable. We could save the examples after every iteration?

**Comment by evg-tyurin** [comment]
Yes, but we probably should save the examples of an iteration separately of other examples/iterations. This would allow to cut older examples in proper time.

**Comment by evg-tyurin** [comment]
I've added 
trainExamplesHistory = []
which holds examples from N latest iterations as well as saving / loading this list to / from file. 

I can open PR after I test it.

**Comment by suragnair** [comment]
Yes let's do this.

**Comment by evg-tyurin** [comment]
Look at #35
